tests/test_utils.py

```diff
@@ -58,3 +66,286 @@ def multiply(a, b):
+from typing import Optional
+from tool import TestTool
+
+tool = TestTool()
+
+launch_gradio_demo(tool)
+"""
+        )
+
+
+def test_e2e_function_tool_save():
+    @tool
+    def test_tool(task: str) -> str:
+        """
+        Test tool description
+
+        Args:
+            task: tool input
+        """
+        import IPython  # noqa: F401
+
+        return task
+
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        test_tool.save(tmp_dir)
+        assert os.listdir(tmp_dir) == ["requirements.txt", "app.py", "tool.py"]
+        assert (
+            pathlib.Path(tmp_dir, "tool.py").read_text()
+            == """from smolagents import Tool
+from typing import Optional
+
+class SimpleTool(Tool):
+    name = "test_tool"
+    description = "Test tool description"
+    inputs = {"task":{"type":"string","description":"tool input"}}
+    output_type = "string"
+
+    def forward(self, task: str) -> str:
+        \"""
+        Test tool description
+
+        Args:
+            task: tool input
+        \"""
+        import IPython  # noqa: F401
+
+        return task"""
+        )
+        requirements = set(pathlib.Path(tmp_dir, "requirements.txt").read_text().split())
+        assert requirements == {"smolagents"}  # FIXME: IPython should be in the requirements
+        assert (
+            pathlib.Path(tmp_dir, "app.py").read_text()
+            == """from smolagents import launch_gradio_demo
+from typing import Optional
+from tool import SimpleTool
+
+tool = SimpleTool()
+
+launch_gradio_demo(tool)
+"""
+        )
+
+
+def test_e2e_ipython_function_tool_save():
+    shell = InteractiveShell.instance()
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        code_blob = textwrap.dedent(f"""
+        from smolagents import tool
+
+        @tool
+        def test_tool(task: str) -> str:
+            \"""
+            Test tool description
+
+            Args:
+                task: tool input
+            \"""
+            import IPython  # noqa: F401
+
+            return task
+
+        test_tool.save("{tmp_dir}")
+        """)
+        assert shell.run_cell(code_blob, store_history=True).success
+        assert os.listdir(tmp_dir) == ["requirements.txt", "app.py", "tool.py"]
+        assert (
+            pathlib.Path(tmp_dir, "tool.py").read_text()
+            == """from smolagents import Tool
+from typing import Optional
+
+class SimpleTool(Tool):
+    name = "test_tool"
+    description = "Test tool description"
+    inputs = {"task":{"type":"string","description":"tool input"}}
+    output_type = "string"
+
+    def forward(self, task: str) -> str:
+        \"""
+        Test tool description
+
+        Args:
+            task: tool input
+        \"""
+        import IPython  # noqa: F401
+
+        return task"""
+        )
+        requirements = set(pathlib.Path(tmp_dir, "requirements.txt").read_text().split())
+        assert requirements == {"smolagents"}  # FIXME: IPython should be in the requirements
```

Comment:
@aymeric-roucher I noticed some differences between tools created by the tool decorator and a Tool subclass, especially the `requirements` here which seem to miss `IPython`. I might miss something!

Reply:
NB: requirements generation is not deterministic, hence the set here